File: week-8-and-9/3-Ruby/fibonacci-sequence/my_solution.rb
# U3.W8-9: 


# I worked on this challenge by myself.

# 2. Pseudocode

<<-Pseudocode

1. if you run num through 5x^2+4 or 5x^2-4 - does it equal a perfect square?
2. i.e. 5x^2+4 ?= ((5x^2+4)^0.5)^2

y^2 === 5x^2+4


Pseudocode

	
# 3. Initial Solution

# def is_fibonacci?(num)

# 	# p (5*(num**2)+4)
# 	# p ((5*(num**2)+4)**0.5).to_i**2

# 	# p (5*(num**2)-4)
# 	# p ((5*(num**2)-4)**0.5).to_i**2


# 	if (5*(num**2)+4) == ((5*(num**2)+4)**0.5).to_i**2 then true
# 	elsif (5*(num**2)-4) == ((5*(num**2)-4)**0.5).to_i**2 then true
# 	else false
# 	end
# end


# is_fibonacci?(0)

# 4. Refactored Solution

def is_fibonacci?(num)
	if (5*(num**2)+4) == ((5*(num**2)+4)**0.5).to_i**2 then true
	elsif (5*(num**2)-4) == ((5*(num**2)-4)**0.5).to_i**2 then true
	else false
	end
end





# 1. DRIVER TESTS GO BELOW THIS LINE

def assert expected, actual, message
	display message
	display actual == expected || display_error(expected, actual)
end

def display message
	p "*"*50
	p message
	p "*"*50
end

def display_error expected, actual
	p "ERROR!  Expected #{expected} but got #{actual}!"
end

assert true, is_fibonacci?(0), "Testing for 0"
assert true, is_fibonacci?(1), "Testing for 1"
assert true, is_fibonacci?(2), "Testing for 2"
assert true, is_fibonacci?(3), "Testing for 3"	
assert false, is_fibonacci?(4), "Testing for 5"


# 5. Reflection 

File: week-8-and-9/2-JavaScript/translate-to-js/my_solution.js
// U3.W8-9: Challenge you're converting - Reverse Words

// I worked on this challenge with Joseph White.

// 2. Pseudocode
// Create function reverse

// reverse accepts string as argument

// reverse takes last character from string and first character
// from string and switches positions

// this process then repeats for 2nd character and 2nd to last character
// etc. until entire string is reversed

// Number of times to switch characters is determined by
// taking length of string/2 rounded down

// the reversed string is then returned


// 3. Initial Solution

function reverse (word) {
	var index = -1
	var reversedWord = ""
	for (var i = 0; i < word.length; i++) {
		reversedWord += word.charAt(word.length + index)
		index -= 1
	}
	return reversedWord
}


// 4. Refactored Solution

function reverse (word) {
	var reversedWord = ""
	for (var i = 0; i < word.length; i++) {
		reversedWord += word.charAt(word.length -1 -i)
	}
	return reversedWord
}

// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

function randomWord (length) {
    var chars = '0123456789abcdefghijklmnopqrstuvwxyz'
    var result = '';
    for (var i = length; i > 0; --i) result += chars[Math.round(Math.random() * (chars.length - 1))];
    return result;
}

function assert(test, message, test_number){
	if(!test){
		console.log(test_number + "false");
		throw "ERROR: " + message;
	}
	console.log (test_number + "true");
	return true;
}

assert(
	(reverse instanceof Function), 
	"The reverse method should be a function.\n", 
	"1. "
)

assert(
	reverse("")==="", 
	"Checking to see if empty string returns empty string",
	"2. "
	)

assert(
	reverse("Thomas")==="samohT", 
	"Checking to see if the opposite of Thomas is samohT",
	"3. "
	)

assert(
	reverse("I like ice cream")==="maerc eci ekil I", 
	"Checking to see if the opposite of 'I like ice cream' is 'maerc eci ekil I'",
	"4. "
	)

// 5. Reflection 
// Relatively easy exercise; Joseph helped think up a more efficient way of tackling the problem than I had thought of before.  Overall - a great session!